man/hfst-format.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH HFST-FORMAT "1" "April 2012" "HFST" "User Commands"
.SH NAME
hfst-format \- =determine HFST transducer format
.SH SYNOPSIS
.B hfst-format
[\fIOPTIONS\fR...] [\fIINFILE\fR]
.SH DESCRIPTION
determine HFST transducer format
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fIINFILE\fR
Read input transducer from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIOUTFILE\fR
Write output transducer to OUTFILE
.SS "Tool-specific options:"
.TP
\fB\-l\fR, \fB\-\-list\-formats\fR
List available transducer formats
and print them to standard output
.TP
\fB\-t\fR, \fB\-\-test\-format\fR FMT
Whether the format FMT is available,
exits with 0 if it is, else with 1
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
Format of result depends on format of INFILE
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://sourceforge.net/tracker/?atid=1061990&group_id=224521&func=browse>
.PP
hfst\-format home page:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstFormat>
.br
General help using HFST software:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstHome>
.br
Transducers in (null) are of type SFST (1.4 compatible)
.SH COPYRIGHT
Copyright \(co 2010 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Transducers in (null) are of type SFST (1.4 compatible)


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.3.9
---------------------------

* Bugs fixed and new features added to hfst-proc

* Xfst rules tentatively supported in hfst-regexp2fst

* hfst-fst2strings does not produce duplicate paths any more 
  when using foma as backend

* hfst-lexc implementation format defaults to openfst-tropical
  instead of foma

Noteworthy changes in 3.3.8
---------------------------

* Fixes 3.3.7

Noteworthy changes in 3.3.7
---------------------------

* Compatible with OpenFst version 1.3.1

* Identity symbols supported in hfst-regexp2fst

Noteworthy changes in 3.3.6
---------------------------

* Include directives of foma and SFST header files changed so
  that standard include path is used instead of source path

* Compilation warnings and some memory leaks fixed

Noteworthy changes in 3.3.5
---------------------------

* hfst-fst2txt.cc also writes to dot/graphviz format

* Epenthesis added for Xerox rules parser

* hfst-compose-intersect gives warnings on missing symbols

Noteworthy changes in 3.3.4
---------------------------

* Substitution functions optimized and new substitution functions added that
  perform multiple substitutions at the same time.

* The tool hfst-substitute now implements the option --in-order.

* Updated to OpenFst version 1.2.10.

* Conversions between transducer formats are more efficient.

Noteworthy changes in 3.3.3
---------------------------

* HfstTransitionGraph and optimized lookup transducer code optimized

* Foma transducer name is not copied to the net but is left empty

* Small fixes to twolc

Noteworthy changes in 3.3.2
---------------------------

* Forward compatibility with OpenFst 1.2.8 is achieved.

* PKG_ macros removed from end user builds since they do not work nicely on
  all platforms.

Noteworthy changes in 3.3.1
---------------------------

* Handling of flag diacritics improved.

* More examples in fsmbook-examples included in tests.

* Extra files created during tests as well as generated autotools files 
  are now ignored by svn.

Noteworthy changes in 3.3.0
---------------------------

* HFST updated to work with the newest versions of backend libraries

* a new replace down function by Karttunen added

* HfstTransitionGraph and conversions between formats made less error-prone

* protected/public changes in API

* SFST and foma libraries are disabled from default build

Noteworthy changes in 3.2.0
---------------------------

* Numerous improvements to hfst-proc, including better uppercasing via glib,
  better compound handling, new analysis mode for raw analysis without case
  changes or special symbols etc.

* old slow lexc (not built by default)

* numerous scripting fixes for mac os x compatibility with foma

* constness changes in API

Noteworthy changes in 3.1.1
---------------------------

* Numerous bugfixes, in particular regarding a broken assumption in
  compose-intersect

Noteworthy changes in 3.1.0
---------------------------

* api for lookup and lookdown changed, previously memory didn't get initialized
  anywhere, causing a serious bug

Noteworthy changes in 3.0.3
---------------------------

* numerous internal improvements, eg. distcheck is now functional

* many bugfixes, especially wrt. INDENTITY_PAIR & OTHER SYMBOL, composition,
  final weights in ol-converted transducers

Noteworthy changes in 3.0.2
---------------------------

* some fixes to OSX compatibility

* no more license conflict problems wrt foma

* better memory use in compose-intersect

* new implementation of internal transducer

Noteworthy changes in 3.0.1
---------------------------

* this is a bugfix release

* conversion to hfst-optimized-lookup format was fundamentally broken in the
  original release; this should provide a fix and considerable speedup

* compose-intersect memory use is much improved

Noteworthy changes in 3.0
-------------------------

* all new API in object oriented form

* brand new foma backend for unweighted transducer handling

* the FST libraries used for automata handling are now dynamically linked to
  HFST

    * this requires end users to install needed backends themselves

* hfst-lexc and hfst-xfst based on foma's compiler

  * xfst is now also fully supported

  * the obsolete command line clients for HFST-based compilation of lexc and
    xfst scripts are hidden in hfst-lexc2fst and hfst-xfst2fst

* preliminary support for unknown character classes

* optimized-lookup backend supported in library

  * dropped special tools for optimised handling since optimised lookup is
    included in library: hfst-optimised-lookup -> hfst-lookup; 
    hfst-lookup-optimize -> hfst-fst2fst

* hfst-proc tool to emulate lt-proc, cg-proc

* Slightly improved string and text formats to support weights uniformly and
  parsing tabs and spaces as automata arcs: see `wiki page for text and string
  formats
  <https://kitwiki.csc.fi/twiki/bin/view/KitWiki/HfstStringAndTextFormats>`

* better error handling using exceptions

* The argument handling order of binary tools changed: stdin is first if not
  specified

* The automata are now packed in HFST automaton container format (akin to AVI
  for videos); this supports arbitrary metadata and underlying FST formats

* the symbol handling is more implicit; no more symbol table arguments for tools
  or hfst-symbols command

* converted to new autotools: silent rules and color tests

* Hfst's internal automata format for fallback on some unsupported operations
  and for conversions

Noteworthy changes in 2.4.1
---------------------------

* Fixed inclusion path in pkg-config file

* Fixed extraneous epsilon path in hfst-lexc transducers

Noteworthy changes in release 2.4
---------------------------------


hfst3/tools/src/hfst-format.cc
//! @file hfst-format.cc
//!
//! @brief Format checking command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using std::ifstream;
using std::ios;

using hfst::HfstTransducer;

static bool list_formats = false;
static char * format_to_test = NULL;

void
print_usage()
{
  // c.f. 
  // http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
  fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
	  "determine HFST transducer format\n"
	  "\n", program_name);
  
  print_common_program_options(message_out);
  print_common_unary_program_options(message_out);
  fprintf
    (message_out, 
     "Tool-specific options:\n"
     "  -l, --list-formats     List available transducer formats\n"
     "                         and print them to standard output\n");
  fprintf
    (message_out, 
     "  -t, --test-format FMT  Whether the format FMT is available,\n"
     "                         exits with 0 if it is, else with 1\n");
  fprintf(message_out, "\n");
  print_common_unary_program_parameter_instructions(message_out);
  fprintf(message_out, "\n");
  print_report_bugs();
  fprintf(message_out, "\n");
  print_more_info();
}


int
parse_options(int argc, char** argv)
{
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
          HFST_GETOPT_COMMON_LONG,
          HFST_GETOPT_UNARY_LONG,
          {"input1", required_argument, 0, '1'},
          {"input2", required_argument, 0, '2'},
          {"list-formats", no_argument, 0, 'l'},
	  {"test-format", required_argument, 0, 't'},
          {0,0,0,0}
        };
        int option_index = 0;
        char c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT "1:2:lt:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }
        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
        case '1':
          inputfilename = strdup(optarg);
          break;
        case '2':
          inputfilename = strdup(optarg);
          break;
	  /*case 'l':
          inputfilename = strdup(optarg);
          break;*/
	case 'l':
	  list_formats=true;
	  break;
	case 't':
	  format_to_test= strdup(optarg);
	  break;
        default:
          // I suppose it's crucial for this tool to ignore other options
          break;
        }
    }

    if (format_to_test != NULL)
      {
	if ((strcmp(format_to_test, "sfst") == 0 &&
	     HfstTransducer::is_implementation_type_available
	     (hfst::SFST_TYPE))
	    ||
	    (strcmp(format_to_test, "openfst-tropical") == 0 &&
	     HfstTransducer::is_implementation_type_available
	     (hfst::TROPICAL_OPENFST_TYPE))
	    ||
	    (strcmp(format_to_test, "openfst-log") == 0 &&
	     HfstTransducer::is_implementation_type_available
	     (hfst::LOG_OPENFST_TYPE))
	    ||
	    (strcmp(format_to_test, "foma") == 0 &&
	     HfstTransducer::is_implementation_type_available
	     (hfst::FOMA_TYPE))
	    ||
	    (strcmp(format_to_test, "optimized-lookup-unweighted") == 0 &&
	     HfstTransducer::is_implementation_type_available
	     (hfst::HFST_OL_TYPE))
	    ||
	    (strcmp(format_to_test, "optimized-lookup-weighted") == 0 &&
	     HfstTransducer::is_implementation_type_available
	     (hfst::HFST_OLW_TYPE))
	    )
	  exit(0);
	exit(1);
	}

    if (list_formats)
      {
	if (HfstTransducer::is_implementation_type_available
	    (hfst::SFST_TYPE))
	  fprintf(stdout, "sfst\n");

	if (HfstTransducer::is_implementation_type_available
	    (hfst::TROPICAL_OPENFST_TYPE))
	  fprintf(stdout, "openfst-tropical\n");
	
	if (HfstTransducer::is_implementation_type_available
	    (hfst::LOG_OPENFST_TYPE))
	  fprintf(stdout, "openfst-log\n");
	
	if (HfstTransducer::is_implementation_type_available
	    (hfst::FOMA_TYPE))
	  fprintf(stdout, "foma\n");
	
	if (HfstTransducer::is_implementation_type_available
	    (hfst::HFST_OL_TYPE))
	  fprintf(stdout, "optimized-lookup-unweighted\n");	      
	
	if (HfstTransducer::is_implementation_type_available
	    (hfst::HFST_OLW_TYPE))
	  fprintf(stdout, "optimized-lookup-weighted\n");
	
	exit(0);
      }

    (void)inputfilename;
    (void)inputNamed;

    try {

      if (inputfilename == NULL)
    {
      if ((argc - optind) == 0)
        {
          inputfilename = strdup("<stdin>");
          hfst::HfstInputStream is("");
          return is.get_type();
        }
      else if ((argc - optind) == 1)
        {
          inputfilename = argv[optind];
        }     
    }
      hfst::HfstInputStream is(inputfilename);
      return is.get_type();
    } // NotTransducerStreamException
    catch (const HfstException e) {
      fprintf(stderr, "ERROR: The file/stream does not contain transducers.\n");
      exit(1);
    }
}


int main (int argc, char * argv[])
{
  hfst_set_program_name(argv[0], "0.1", "HfstFormat");
  verbose = true;
  hfst::ImplementationType type = static_cast<hfst::ImplementationType>(parse_options(argc,argv));
  verbose_printf("Transducers in %s are of type %s\n", inputfilename,
                 hfst_strformat(type));
}
